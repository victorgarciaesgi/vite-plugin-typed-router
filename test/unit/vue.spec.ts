import { readFile } from 'fs/promises';

describe('Init route file', () => {
  it('should generate to correct tree for the pages folder', async () => {
    const routeFile = await readFile('playground/vue/src/generated/__routes.ts', 'utf-8');
    const definitionsFile = await readFile(
      'playground/vue/src/generated/typed-router.d.ts',
      'utf-8'
    );
    const hookFile = await readFile('playground/vue/src/generated/__useTypedRouter.ts', 'utf-8');
    expect(routeFile).toMatchSnapshot();
    expect(definitionsFile).toMatchSnapshot();
    expect(hookFile).toMatchSnapshot();
  });
});
