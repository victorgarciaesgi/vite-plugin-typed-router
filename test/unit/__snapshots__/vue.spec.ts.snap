// Vitest Snapshot v1

exports[`Init route file > should generate to correct tree for the pages folder 1`] = `
"/**
 * ---------------------
 * ðŸš—ðŸš¦ Generated by vite-plugin-vue-router-typed. Do not modify !
 * ---------------------
 * */

export const routesNames = {
  activate: 'activate' as const,
  index: 'index' as const,
  childOne: {
    childOneChildOneSubOne: 'parent-child-one-child-one-sub-one' as const,
    user: { oneChildOneSubOneUserId: 'parent-child-one-child-one-sub-one-user-id' as const },
    childOneChildOneSubTwo: 'parent-child-one-child-one-sub-two' as const,
    index: 'parent-child-one' as const,
  },
  childTwo: {
    childTwoId: 'parent-child-two-id' as const,
    childTwoChildOneSubOneUser: 'parent-child-two-child-one-sub-one-user' as const,
    childTwoChildOneSubOne: 'parent-child-two-child-one-sub-one' as const,
    index: 'parent-child-two' as const,
    profile: {
      id: {
        profileIdSlug: 'parent-child-two-profile-id-slug' as const,
        index: 'parent-child-two-profile-id' as const,
      },
      index: 'parent-child-two-profile' as const,
    },
  },
  rootPage: 'rootPage' as const,
  test: 'test' as const,
};

export type TypedRouteList =
  | 'activate'
  | 'index'
  | 'parent-child-one-child-one-sub-one'
  | 'parent-child-one-child-one-sub-one-user-id'
  | 'parent-child-one-child-one-sub-two'
  | 'parent-child-one'
  | 'parent-child-two-id'
  | 'parent-child-two-child-one-sub-one-user'
  | 'parent-child-two-child-one-sub-one'
  | 'parent-child-two'
  | 'parent-child-two-profile-id-slug'
  | 'parent-child-two-profile-id'
  | 'parent-child-two-profile'
  | 'rootPage'
  | 'test';
"
`;

exports[`Init route file > should generate to correct tree for the pages folder 2`] = `
"/**
 * ---------------------
 * ðŸš—ðŸš¦ Generated by vite-plugin-vue-router-typed. Do not modify !
 * ---------------------
 * */

import type {
  NavigationFailure,
  RouteLocation,
  RouteLocationNormalizedLoaded,
  RouteLocationOptions,
  RouteQueryAndHash,
} from 'vue-router';
import type { TypedRouteList } from './__routes';

export type RouteListDecl = {
  activate: 'activate';
  index: 'index';
  childOne: {
    childOneChildOneSubOne: 'parent-child-one-child-one-sub-one';
    user: { oneChildOneSubOneUserId: 'parent-child-one-child-one-sub-one-user-id' };
    childOneChildOneSubTwo: 'parent-child-one-child-one-sub-two';
    index: 'parent-child-one';
  };
  childTwo: {
    childTwoId: 'parent-child-two-id';
    childTwoChildOneSubOneUser: 'parent-child-two-child-one-sub-one-user';
    childTwoChildOneSubOne: 'parent-child-two-child-one-sub-one';
    index: 'parent-child-two';
    profile: {
      id: {
        profileIdSlug: 'parent-child-two-profile-id-slug';
        index: 'parent-child-two-profile-id';
      };
      index: 'parent-child-two-profile';
    };
  };
  rootPage: 'rootPage';
  test: 'test';
};

export type TypedRouteParams = {
  activate: never;
  index: never;
  'parent-child-one-child-one-sub-one': never;
  'parent-child-one-child-one-sub-one-user-id': {
    id: string | number;
  };
  'parent-child-one-child-one-sub-two': never;
  'parent-child-one': never;
  'parent-child-two-id': {
    id: string | number;
  };
  'parent-child-two-child-one-sub-one-user': {
    user?: string | number;
  };
  'parent-child-two-child-one-sub-one': never;
  'parent-child-two': never;
  'parent-child-two-profile-id-slug': {
    slug: string | number;
    id?: string | number;
  };
  'parent-child-two-profile-id': {
    id: string | number;
  };
  'parent-child-two-profile': never;
  rootPage: never;
  test: never;
};

type TypedRouteParamsStructure = {
  [K in TypedRouteList]: Record<string, string | number> | never;
};

type AreAllParamsOptional<T> = {
  [K in keyof T]-?: T[K] extends Exclude<T[K], undefined> ? false : true;
}[keyof T] extends true
  ? true
  : false;

type TypedLocationAsRelativeRaw<T extends TypedRouteList, V extends any = TypedRouteParams[T]> = {
  name?: T;
} & ([V] extends [never]
  ? {}
  : AreAllParamsOptional<V> extends true
  ? { params?: V }
  : { params: V });

type TypedRouteLocationRaw<T extends TypedRouteList> = RouteQueryAndHash &
  TypedLocationAsRelativeRaw<T> &
  RouteLocationOptions;

export interface TypedRouter {
  /**
   * Remove an existing route by its name.
   *
   * @param name - Name of the route to remove
   */
  removeRoute(name: TypedRouteList): void;
  /**
   * Checks if a route with a given name exists
   *
   * @param name - Name of the route to check
   */
  hasRoute(name: TypedRouteList): boolean;
  /**
   * Returns the {@link RouteLocation | normalized version} of a
   * {@link RouteLocationRaw | route location}. Also includes an \`href\` property
   * that includes any existing \`base\`. By default the \`currentLocation\` used is
   * \`route.currentRoute\` and should only be overriden in advanced use cases.
   *
   * @param to - Raw route location to resolve
   * @param currentLocation - Optional current location to resolve against
   */
  resolve<T extends TypedRouteList>(
    to: TypedRouteLocationRaw<T>,
    currentLocation?: RouteLocationNormalizedLoaded
  ): RouteLocation & {
    href: string;
  };
  /**
   * Programmatically navigate to a new URL by pushing an entry in the history
   * stack.
   *
   * @param to - Route location to navigate to
   */
  push<T extends TypedRouteList>(
    to: TypedRouteLocationRaw<T>
  ): Promise<NavigationFailure | void | undefined>;
  /**
   * Programmatically navigate to a new URL by replacing the current entry in
   * the history stack.
   *
   * @param to - Route location to navigate to
   */
  replace<T extends TypedRouteList>(
    to: TypedRouteLocationRaw<T>
  ): Promise<NavigationFailure | void | undefined>;
}
"
`;

exports[`Init route file > should generate to correct tree for the pages folder 3`] = `
"/**
 * ---------------------
 * ðŸš—ðŸš¦ Generated by vite-plugin-vue-router-typed. Do not modify !
 * ---------------------
 * */

import { useRouter } from 'vue-router';
import { TypedRouter, RouteListDecl } from './typed-router';

/** Returns instances of $typedRouter and $routesList fully typed to use in your components or your Vuex/Pinia store
 *
 * @exemple
 *
 * \`\`\`ts
 * const { router, routes } = useTypedRouter();
 * \`\`\`
 */
export const useTypedRouter = (): {
  /** Export of $router with type check */
  router: TypedRouter;
  /** Contains a typed dictionnary of all your route names (for syntax sugar) */
  routes: RouteListDecl;
} => {
  const router = useRouter();

  const routesList = {
    activate: 'activate',
    index: 'index',
    childOne: {
      childOneChildOneSubOne: 'parent-child-one-child-one-sub-one',
      user: { oneChildOneSubOneUserId: 'parent-child-one-child-one-sub-one-user-id' },
      childOneChildOneSubTwo: 'parent-child-one-child-one-sub-two',
      index: 'parent-child-one',
    },
    childTwo: {
      childTwoId: 'parent-child-two-id',
      childTwoChildOneSubOneUser: 'parent-child-two-child-one-sub-one-user',
      childTwoChildOneSubOne: 'parent-child-two-child-one-sub-one',
      index: 'parent-child-two',
      profile: {
        id: {
          profileIdSlug: 'parent-child-two-profile-id-slug',
          index: 'parent-child-two-profile-id',
        },
        index: 'parent-child-two-profile',
      },
    },
    rootPage: 'rootPage',
    test: 'test',
  };

  return {
    router: router,
    routes: routesList,
  } as any;
};
"
`;
